File: app/src/main/java/com/example/desiregallery/network/DGNetwork.java
package com.example.desiregallery.network;

import com.example.desiregallery.models.Post;
import com.example.desiregallery.models.User;
import com.example.desiregallery.network.serializers.*;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import java.lang.reflect.Type;
import java.util.List;

public class DGNetwork {
    private static final String BASE_API = "https://firestore.googleapis.com/v1/projects/desiregallery-8072a/databases/(default)/documents/";

    private static IDGApi service;
    private static Gson gson;

    public static IDGApi getService() {
        if (service == null)
            initService();
        return service;
    }

    private static Gson getGson() {
        if (gson == null) {
            Type postListType = new TypeToken<List<Post>>() {}.getType();
            Type userListType = new TypeToken<List<User>>() {}.getType();
            GsonBuilder gsonBuilder = new GsonBuilder()
                    .registerTypeAdapter(Post.class, new PostDeserializer())
                    .registerTypeAdapter(postListType, new PostListDeserializer())
                    .registerTypeAdapter(User.class, new UserDeserializer())
                    .registerTypeAdapter(User.class, new UserSerializer())
                    .registerTypeAdapter(userListType, new UserListDeserializer());
            gson = gsonBuilder.create();
        }
        return gson;
    }

    private static void initService() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_API)
                .addConverterFactory(GsonConverterFactory.create(getGson()))
                .build();
        service = retrofit.create(IDGApi.class);
    }
}

File: app/src/main/java/com/example/desiregallery/network/DGNetwork.kt
@file:JvmName("DGNetwork")
package com.example.desiregallery.network

import com.example.desiregallery.logging.logInfo
import com.example.desiregallery.models.User
import retrofit2.Retrofit
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

private const val TAG = "DGNetwork"
private const val BASE_URL = "https://firestore.googleapis.com/v1/projects/desiregallery-8072a/databases/(default)/documents/"
private const val QUERY_URL = "https://firestore.googleapis.com/v1/projects/desiregallery-8072a/databases/(default)/documents:runQuery/"

val baseService: IDGApi by lazy {
    val retrofit = Retrofit.Builder()
        .baseUrl(BASE_URL)
        .addConverterFactory(createBaseGson())
        .build()
    retrofit.create(IDGApi::class.java)
}
val queryService: IQueryApi  by lazy {
    val retrofit = Retrofit.Builder()
        .baseUrl(QUERY_URL)
        .addConverterFactory(createQueryGson())
        .build()
    retrofit.create(IQueryApi::class.java)
}

fun getUsersByNames(userNames: Collection<String>): List<User> {
    logInfo(TAG, "Preparing to load ${userNames.size} users")
    val users = ArrayList<User>()
    val executor = Executors.newCachedThreadPool()
    for (name in userNames)
        executor.execute {
            val userResponse = baseService.getUser(name).execute()
            if (!userResponse.isSuccessful || userResponse.body() == null)
                throw Exception("There was an error while fetching authors")

            users.add(userResponse.body()!!)
        }
    executor.shutdown()
    executor.awaitTermination(15, TimeUnit.SECONDS)
    logInfo(TAG, "Loaded ${users.size} users")
    if (users.size != userNames.size)
        throw Exception("There was an error while fetching authors")
    return users
}
